crates/runtime/src/task_history.rs

```diff
@@ -0,0 +1,377 @@
+    ($field:expr, $field_name:expr, $missing_fields:expr) => {
+        if $field.is_none() {
+            $missing_fields.push($field_name.into());
+        }
+    };
+}
+
+pub enum TaskType {
+    Query,
+    Chat,
+    Embed,
+    Search,
+    FunctionCall,
+}
+
+impl Display for TaskType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            TaskType::Query => write!(f, "query"),
+            TaskType::Chat => write!(f, "chat"),
+            TaskType::Embed => write!(f, "embed"),
+            TaskType::Search => write!(f, "search"),
+            TaskType::FunctionCall => write!(f, "function_call"),
+        }
+    }
+}
+
+fn convert_hashmap_to_maparray(labels: &HashMap<String, String>) -> MapArray {
+    let keys_field = Arc::new(Field::new("keys", DataType::Utf8, false));
+    let values_field = Arc::new(Field::new("values", DataType::Utf8, false));
+
+    let (keys, values): (Vec<&str>, Vec<&str>) =
+        labels.iter().map(|(k, v)| (k.as_str(), v.as_str())).unzip();
+
+    let keys_array = StringArray::from(keys);
+    let values_array = StringArray::from(values);
+
+    let entry_struct = StructArray::from(vec![
+        (
+            keys_field.clone(),
+            Arc::new(keys_array) as Arc<dyn arrow::array::Array>,
+        ),
+        (
+            values_field.clone(),
+            Arc::new(values_array) as Arc<dyn arrow::array::Array>,
+        ),
+    ]);
+
+    let entry_offsets = Buffer::from_vec(vec![0, labels.len() as i32]);
+    let map_data_type = DataType::Map(
+        Arc::new(Field::new_struct(
+            "entries",
+            vec![
+                Arc::new(Field::new("keys", DataType::Utf8, false)),
+                Arc::new(Field::new("values", DataType::Utf8, false)),
+            ],
+            false,
+        )),
+        false,
+    );
+
+    let map_data = ArrayData::builder(map_data_type)
+        .len(1)
+        .add_buffer(entry_offsets)
+        .add_child_data(entry_struct.to_data())
+        .build()
+        .unwrap();
+
+    MapArray::from(map_data)
+}
+
+pub(crate) struct TaskTracker {
+    pub(crate) df: Arc<crate::datafusion::DataFusion>,
+    pub(crate) id: Uuid,
+    pub(crate) context_id: Uuid,
+    pub(crate) parent_id: Option<Uuid>,
+
+    pub(crate) task_type: TaskType,
+    pub(crate) input_text: Arc<str>,
+
+    pub(crate) start_time: SystemTime,
+    pub(crate) end_time: Option<SystemTime>,
+    pub(crate) execution_time: Option<f32>,
+    pub(crate) outputs_produced: u64,
+    pub(crate) cache_hit: Option<bool>,
+    pub(crate) error_message: Option<String>,
+    pub(crate) labels: HashMap<String, String>,
+
+    pub(crate) timer: Instant,
+}
+
+impl TaskTracker {
+    pub fn new(
+        df: Arc<crate::datafusion::DataFusion>,
+        context_id: Uuid,
+        task_type: TaskType,
+        input_text: Arc<str>,
+        id: Option<Uuid>,
+        labels: Option<HashMap<String, String>>,
```

Comment:
Anything optional should be specified by functions via a builder-like pattern.

Reply:
I will change labels. I will keep `id: Option<Uuid>`. The intention is that None -> autogenerate.